rfcs/0001-messages.md

```diff
@@ -0,0 +1,423 @@
+
+## Definitions
+
+- `dealer` - Participant who distributes the initial package to all the other participants.
+- `aggregator` - Participant in charge of collecting all the signatures from the other participants and generating the final group signature.
+- `signer` - Participant that will receive the initial package, sign and send the signature to the aggregator to receive the final group signature.
+
+Note: In this RFC we consider the above 3 participants to be different. `dealer` and `aggregator` have specific hard coded `ParticipantId`s, so for example a `dealer` can't be a `signer`. This is not a protocol limitation but a specific rule introduced in this document.
+
+## Guide-level explanation
+
+We propose a message separated in 2 parts, a header and a payload:
+
+```rust
+/// The data required to serialize a frost message.
+struct Message {
+    header: Header,
+    payload: Payload,
+}
+```
+
+`Header` will look as follows:
+
+```rust
+/// The data required to serialize the common header fields for every message.
+///
+/// Note: the `msg_type` is derived from the `payload` enum variant.
+struct Header {
+    version: MsgVersion,
+    sender: ParticipantId,
+    receiver: ParticipantId,
+}
+```
+
+While `Payload` will be defined as:
+
+```rust
+/// The data required to serialize the payload for a message.
+enum Payload {
+    SharePackage(messages::SharePackage),
+    SigningCommitments(messages::SigningCommitments),
+    SigningPackage(messages::SigningPackage),
+    SignatureShare(messages::SignatureShare),
+    AggregateSignature(messages::AggregateSignature),
+}
+```
+
+All the messages and new types will be defined in a new file `src/frost/messages.rs`
+
+## Reference-level explanation
+
+Here we explore in detail the header types and all the message payloads. 
+
+### Header
+
+Fields of the header define new types. Proposed implementation for them is as follows:
+
+```rust
+/// The numeric values used to identify each `Payload` variant during serialization.
+#[repr(u8)]
+#[non_exhaustive]
+enum MsgType {
+    SharePackage,
+    SigningCommitments,
+    SigningPackage,
+    SignatureShare,
+    AggregateSignature,
+}
+
+/// The numeric values used to identify the protocol version during serialization.
+struct MsgVersion(u8);
+
+const BASIC_FROST_SERIALIZATION: MsgVersion = MsgVersion(0);
+
+/// The numeric values used to identify each participant during serialization.
+///
+/// In the `frost` module, participant ID `0` should be invalid.
+/// But in serialization, we want participants to be indexed from `0..n`,
+/// where `n` is the number of participants.
+/// This helps us look up their shares and commitments in serialized arrays.
+/// So in serialization, we assign the dealer and aggregator the highest IDs,
+/// and mark those IDs as invalid for signers.
```

Comment:
This specification change allows us to adapt to discontinuous FROST IDs.

It won't change any code now - only when we change FROST ID generation.
```suggestion
/// So in serialization, we assign the dealer and aggregator the highest IDs,
/// and mark those IDs as invalid for signers. Then we serialize the
/// participants in numeric order of their FROST IDs.
```